src/lib/Inference.hs

```diff
@@ -33,12 +33,17 @@ data RequiredTy a = Check a | Infer
 
 inferModule :: TopEnv -> UModule -> Except Module
 inferModule scope (UModule decls) = do
-  ((), (_, decls')) <- runUInferM (mapM_ (inferUDecl True) decls) mempty scope
-  return $ Module Typed decls' mempty
+  ((), (bindings, decls')) <- runUInferM mempty scope $
+                                mapM_ (inferUDecl True) decls
+  let bindings' = envFilter bindings $ \(_, b) -> case b of
+                    DataBoundTypeCon _   -> True
+                    DataBoundDataCon _ _ -> True
+                    _ -> False
+  return $ Module Typed decls' bindings'
```

Comment:
Do I understand correctly that `bindings` contains the local scope, which is where the ADT declarations end up? Why don't we declare them as globals and put them together with `decls`?

Reply:
This way the core IR doesn't need syntax for data definitions at all. They don't really represent "computation", so that seems appropriate to me.